lrpar/src/lib/ctbuilder.rs

```diff
@@ -1209,20 +1209,27 @@ where
     }
 
     /// Generate the user action functions (if any).
-    fn gen_user_actions(&self, grm: &YaccGrammar<StorageT>) -> Result<String, Box<dyn Error>> {
-        let mut outs = String::new();
-
-        if let Some(s) = grm.programs() {
-            outs.push_str("\n// User code from the program section\n\n");
-            outs.push_str(s);
-            outs.push_str("\n// End of user code from the program section\n\n");
-        }
-
+    fn gen_user_actions(&self, grm: &YaccGrammar<StorageT>) -> Result<TokenStream, Box<dyn Error>> {
+        let programs = grm
+            .programs()
+            .as_ref()
+            .map(|s| str::parse::<TokenStream>(s))
+            .transpose()?;
+        let mut action_fns = TokenStream::new();
         // Convert actions to functions
-        outs.push_str("\n    // User actions\n\n");
-        let (parse_paramname, parse_paramdef) = match grm.parse_param() {
-            Some((name, tyname)) => (name.to_owned(), format!("{}: {}", name, tyname)),
-            None => ("()".to_owned(), "_: ()".to_owned()),
+        let (parse_paramname, parse_paramdef, parse_param_unit);
+        match grm.parse_param() {
+            Some((name, tyname)) => {
+                parse_param_unit = tyname.trim() == "()";
```

Comment:
pedantic, but being user provided this can potentially be "( )", with a space in the between parenthesis,
so this probably better stripping all whitespace.  I suppose the existing code suffered the same problem though.

Not much of a problem really, the worst that happens is the generated code emits a warning.
I'm inclined to just leave this for a future cleanup, to avoid introducing changes in this refactor.

Reply:
I also thought the same when I reviewed: I wondered if there was a neater way of asking "is this the unit type", but short of "replace all spaces with ''" I couldn't think of much! I agree it can be done in a future PR with additional tests.